core_matcher.go

```diff
@@ -70,14 +72,22 @@ func (m *coreMatcher) addPattern(x X, patternJSON string) error {
 	freshStart := &coreStart{}
 	freshStart.namesUsed = make(map[string]bool)
 	current := m.start()
+	freshStart.pathsIndex = current.pathsIndex.copy() // NOTE: not sure if this is thread-safe
```

Comment:
Hmm, looks thread-safe to me but there's any doubt should do concurrency fuzz-testing.

Reply:
It looks thread-safe to me as well, but I wasn't sure what happens if during the `copy` another go-routine changes the `nodes` / `fields` (important to note we have only additions, not deletions). The problem with that in theory that we might loose changes, since we are working with specific snapshot.

I'll do concurrency fuzz-testing, never did one, but will look at `concurrency_tests` as a guidance.

Reply:
But, the event body, and then the nodes and fields, are only visible to the thread that's running `matchesForSortedFields`, right?

Reply:
Yep you are right, After thinking about this again, I don't see any issue with thread-safety here. removing this comment.